framework/src/preconditioners/DualMortarPreconditioner.C

```diff
@@ -0,0 +1,588 @@
+    for(auto it=var.active_subdomains().begin(); it != var.active_subdomains().end(); ++it)
+    {
+      SubdomainID sd=(SubdomainID)(*it);
+      std::set<dof_id_type> dofs=_dof_sets[vn][sd];
+      // remove dofs on the interface
+      for (auto it_primary = _dof_sets_primary[vn].begin(); it_primary!=_dof_sets_primary[vn].end(); ++it_primary)
+        dofs.erase(*it_primary);
+      for (auto it_secondary = _dof_sets_secondary[vn].begin(); it_secondary!=_dof_sets_secondary[vn].end(); ++it_secondary)
+        dofs.erase(*it_secondary);
+
+      std::vector<dof_id_type> vec_dofs(dofs.begin(), dofs.end());
+      _dof_sets_interior[vn].insert(make_pair(sd, vec_dofs));
+    }
+  }
+}
+
+void
+DualMortarPreconditioner::getDofVarInterface()
+{
+  _dof_sets_primary.resize(_n_vars);
+  _dof_sets_secondary.resize(_n_vars);
+  for (unsigned int vn = 0; vn < _n_vars; vn++)
+  {
+    // loop over boundary nodes
+    ConstBndNodeRange & range = * _mesh->getBoundaryNodeRange();
+    std::vector<dof_id_type> di;
+    for (const auto & bnode : range)
+    {
+      const Node * node_bdry = bnode->_node;
+      BoundaryID boundary_id = bnode->_bnd_id;
+
+      if (boundary_id == _secondary_boundary)
+      {
+          _dofmap->dof_indices(node_bdry, di, vn);
+          for (auto it = di.begin(); it!=di.end(); ++it)
+            if(std::find(_dof_sets_secondary[vn].begin(), _dof_sets_secondary[vn].end(), *it)==_dof_sets_secondary[vn].end())
+              _dof_sets_secondary[vn].push_back(*it);
+      }
+
+      if (boundary_id == _primary_boundary)
+      {
+        _dofmap->dof_indices(node_bdry, di, vn);
+        for (auto it = di.begin(); it!=di.end(); ++it)
+          if(std::find(_dof_sets_primary[vn].begin(), _dof_sets_primary[vn].end(), *it)==_dof_sets_primary[vn].end())
+            _dof_sets_primary[vn].push_back(*it);
+      }
+    }
+  }
+}
+
+void
+DualMortarPreconditioner::printDofSets()
+{
+  /// Dofs in the interior
+  for (auto it1 = _dof_sets_interior.begin(); it1 != _dof_sets_interior.end(); it1++)
+  {
+    std::cout << "variable "<<std::endl;
+    for (auto it2 = it1->begin(); it2 != it1->end(); it2++)
+    {
+      std::cout << " \tdofs in subdomain " << it2->first<<": ";
+      for (auto it3 = it2->second.begin(); it3 != it2->second.end(); it3++)
+        std::cout <<*it3 << " ";
+      std::cout <<"\n";
+    }
+  }
+
+  // primary
+  std::cout<<"primary dofs:\n";
+  for (auto it1=_dof_sets_primary.begin(); it1!=_dof_sets_primary.end(); it1++)
+  {
+    std::cout<< "\tvariable :";
+    for (auto it2=it1->begin(); it2!=it1->end(); it2++)
+      std::cout<<*it2<<" ";
+    std::cout<<std::endl;
+  }
+
+  // secondary
+  std::cout<<"secondary dofs:\n";
+  for (auto it1=_dof_sets_secondary.begin(); it1!=_dof_sets_secondary.end(); it1++)
+  {
+    std::cout<< "\tvariable :";
+    for (auto it2=it1->begin(); it2!=it1->end(); it2++)
+      std::cout<<*it2<<" ";
+    std::cout<<std::endl;
+  }
+}
+
+void
```

Comment:
![](https://user-images.githubusercontent.com/58825741/86960979-a4b8ac80-c11d-11ea-989c-10ba20962ea8.png)

Reply:
I need this picture in order to understand the code

Reply:
My apologies for the confusion. In the code I use "1" as the primal and "2" as the secondary... Just to match the naming. While in the figure it is the opposite. I will update the figure for your easier understanding

Reply:
Please refer to this figure instead @fdkong :
![Screen Shot 2020-07-10 at 12 16 19 PM](https://user-images.githubusercontent.com/58825741/87185604-4b7c8480-c2a7-11ea-94b2-b5f4f2a10b7f.png)

Reply:
Cool! It is really helpful.